src/Cryptol/Eval/Concrete/Value.hs

```diff
@@ -0,0 +1,353 @@
+  wordMinus _ (BV i x) (BV j y)
+    | i == j = pure $! mkBv i (x-y)
+    | otherwise = panic "Attempt to subtract words of different sizes: wordMinus" [show i, show j]
+
+  wordMult _ (BV i x) (BV j y)
+    | i == j = pure $! mkBv i (x*y)
+    | otherwise = panic "Attempt to multiply words of different sizes: wordMult" [show i, show j]
+
+  wordDiv sym (BV i x) (BV j y)
+    | i == 0 && j == 0 = pure $! mkBv 0 0
+    | i == j =
+        do assertSideCondition sym (y /= 0) DivideByZero
+           pure $! mkBv i (x `div` y)
+    | otherwise = panic "Attempt to divide words of different sizes: wordDiv" [show i, show j]
+
+  wordMod sym (BV i x) (BV j y)
+    | i == 0 && j == 0 = pure $! mkBv 0 0
+    | i == j =
+        do assertSideCondition sym (y /= 0) DivideByZero
+           pure $! mkBv i (x `mod` y)
+    | otherwise = panic "Attempt to mod words of different sizes: wordMod" [show i, show j]
+
+  wordSignedDiv sym (BV i x) (BV j y)
+    | i == 0 && j == 0 = pure $! mkBv 0 0
+    | i == j =
+        do assertSideCondition sym (y /= 0) DivideByZero
+           let sx = signedValue i x
+               sy = signedValue i y
+           pure $! mkBv i (sx `quot` sy)
+    | otherwise = panic "Attempt to divide words of different sizes: wordSignedDiv" [show i, show j]
+
+  wordSignedMod sym (BV i x) (BV j y)
+    | i == 0 && j == 0 = pure $! mkBv 0 0
+    | i == j =
+        do assertSideCondition sym (y /= 0) DivideByZero
+           let sx = signedValue i x
+               sy = signedValue i y
+           pure $! mkBv i (sx `rem` sy)
+    | otherwise = panic "Attempt to mod words of different sizes: wordSignedMod" [show i, show j]
+
+  wordExp _ (BV i x) (BV j y)
+    | i == 0 && j == 0 = pure $! mkBv 0 0
+    | i == j =
+        do let modulus = 0 `setBit` fromInteger i
+           pure . mkBv i $! doubleAndAdd x y modulus
+    | otherwise = panic "Attempt to exp words of different sizes: wordSignedMod" [show i, show j]
+
+  wordLg2 _ (BV i x) = pure $! mkBv i (lg2 x)
+
+  intEq _ x y = pure $! x == y
+  intLessThan _ x y = pure $! x < y
+  intGreaterThan _ x y = pure $! x > y
+
+  intPlus  _ x y = pure $! x + y
+  intMinus _ x y = pure $! x - y
+  intNegate _ x  = pure $! negate x
+  intMult  _ x y = pure $! x * y
+  intDiv sym x y =
+    do assertSideCondition sym (y /= 0) DivideByZero
+       pure $! x `div` y
+  intMod sym x y =
+    do assertSideCondition sym (y /= 0) DivideByZero
+       pure $! x `mod` y
+  intExp sym x y =
+    do assertSideCondition sym (y >= 0) NegativeExponent
+       pure $! x ^ y
+  intLg2 sym x =
+    do assertSideCondition sym (x >= 0) LogNegative
+       pure $! lg2 x
+
+  intToZn _ 0 x = pure x
+  intToZn _ m x = pure $! x `mod` m
+
+  -- NB: requires we maintain the invariant that
+  --     Z_n is in reduced form
+  znToInt _ _m x = pure x
+  znEq _ _m x y = pure $! x == y
+
+  znPlus  _ = liftBinIntMod (+)
+  znMinus _ = liftBinIntMod (-)
+  znMult  _ = liftBinIntMod (*)
+  znNegate _ 0 x = pure $! (negate x)
```

Comment:
This should probably be a panic. Any use of `Z n` implies the constraint `n >= 1`, so modulus of 0 should never happen.

Reply:
OK, makes sense